flink-table/flink-table-common/src/main/java/org/apache/flink/table/sources/TableSourceValidation.java

```diff
@@ -181,6 +182,16 @@ private static void validateLogicalTypeEqualsPhysical(
 			TableSource<?> tableSource) {
 		ResolvedField resolvedField = resolveField(fieldName, tableSource);
 		if (!resolvedField.getType().equals(logicalType)) {
+
+			if (resolvedField.getType().getLogicalType() instanceof LegacyTypeInformationType &&
+				logicalType.getLogicalType().getTypeRoot() == LogicalTypeRoot.DECIMAL &&
+				logicalType.getLogicalType().getTypeRoot() == resolvedField.getType().getLogicalType().getTypeRoot() &&
+				logicalType.getConversionClass() == resolvedField.getType().getConversionClass()) {
```

Comment:
Is that logical equal is enough?

Reply:
The logical types are not equal, one is `DecimalType`, the other is `LegacyTypeInformationType`.

Reply:
I mean above `if (!resolvedField.getType().equals(logicalType))` should be `if (!resolvedField.getType().getLogicalType().equals(logicalType.getLogicalType()))`.

Reply:
And in the case of decimal, we don't need compare `ConversionClass`.

Reply:
I think you are right. We only need to compare `LogicalType` here. But for the decimal, we still need the special condition, but remove the conversion class condition.